R/fast_forward.R

```diff
@@ -0,0 +1,623 @@
+    new_csv
+  }
+
+  csv <- expand_semicolons(csv, "modelling_group")
+  csv <- expand_semicolons(csv, "scenario")
+
+  # This is a little clumsy, but if there is a modelling_group '*',
+  # expand this so we have a line with a named modelling_group
+  # for each that already has a
+  # responsibility_set in the origin touchstone.
+
+  while (any(csv$modelling_group == "*")) {
+    i <- which(csv$modelling_group == "*")[1]
+    row <- csv[i, ]
+    mgs <- DBI::dbGetQuery(con, "
+      SELECT DISTINCT modelling_group
+        FROM responsibility_set
+       WHERE touchstone = $1",
+          csv$touchstone_from[i])$modelling_group
+
+    csv <- csv[-i, ]
+    for (mg in mgs) {
+      row$modelling_group <- mg
+      csv <- rbind(csv, row)
+    }
+  }
+
+  # Having done all the expansion of modelling groups,
+  # all items in that column should now exist.
+
+  mgs <- missing_things(unique(csv$modelling_group), "modelling_group", con)
+  if (length(mgs) > 0) {
+    stop(sprintf("Modelling group(s) not found: %s",
+                 paste(mgs, collapse = ", ")))
+  }
+
+  # We also allow wildcard for scenarios...
+  # Here expand that now ito separate rows for individual scenarios,
+  # that the (already single) modelling group has in
+  # the origin touchstone.
+
+  while (any(csv$scenario == "*")) {
+    i <- which(csv$scenario == "*")[1]
+    row <- csv[i, ]
+    scenarios <- DBI::dbGetQuery(con, "
+      SELECT DISTINCT scenario_description
+        FROM scenario
+        JOIN responsibility
+          ON scenario.id = responsibility.scenario
+        JOIN responsibility_set
+          ON responsibility.responsibility_set = responsibility_set.id
+       WHERE responsibility_set.modelling_group = $1
+         AND scenario.touchstone = $2",
+    list(csv$modelling_group[i], csv$touchstone_from[i]))$scenario_description
+    csv <- csv[-i, ]
+    for (scenario in scenarios) {
+      row$scenario <- scenario
+      csv <- rbind(csv, row)
+    }
+  }
+
+  # Having expanded scenarios to one per line, check they
+  # all exist - noting we are keeping these human readable,
+  # so are really scenario.scenario_description
+
+  scs <- missing_things(unique(csv$scenario), "scenario", con,
+                        "scenario_description")
+  if (length(scs) > 0) {
+    stop(sprintf("Scenario(s) not found: %s",
+                    paste(scs, collapse = ", ")))
+  }
+
+  # So now we have a list of jobs to do:-
+  #
+  # modelling_group, touchstone_from, touchstone_to, scenario
+
+  # We'll remove any invalid combos. Build a hash of
+  # touchstone_from \r modelling_group \r scenario and
+  # touchstone_to \r modelling_group \r scenario to see
+  # what exists in the database.
+
+  # We'll drop anything in the CSV that doesn't exist in the origin
+  # touchstone, as there's nothing to fast-forward. But let's not
+  # make that an error, otherwise we'll get lots of false errors
+  # if we expand (*, *)
+
+  # burden_estimate_set exists for that combination,
```

Comment:
is there a line missing from the comments here? these last 2 lines don't quite scan

Reply:
80-column has worked against me! 

The false errors will occur if we expand (*,*) - I don't want to be strict about the particular (a,b) pairs existing - I'm happy to select all the (a) and all the (b), and only do any work on the valid (a,b)